### PR TITLE
Interact Objects fixes

### PR DESCRIPTION
**Pull request by Zintixx**
- Parsing `interactobject.xml` for details of objects
- Parsing all types of interact objects. All interact objects should now be loaded
- Cleanup/fixes on recipe parser

### CODE REVIEW

File: GameDataParser/Parsers/MapEntityParser.cs [hunk 1]
```diff
@@ -420,5 +410,15 @@ private void BuildMetadata(string xblock, IEnumerable<IMapEntity> mapEntities)
...
             Entities.Add(metadata);
         }
+
+        private InteractObjectType GetInteractObjectType(int interactId)
+        {
+            if (InteractTypes.ContainsKey(interactId))
+            {
+                return InteractTypes[interactId];
+            }
+            Console.WriteLine($"Did not find InteractId: {interactId}");
```
Comment (Sparkymod) [inline GameDataParser/Parsers/MapEntityParser.cs:420]: Debug ?

File: MapleServer2/Types/InteractObject.cs [hunk 2]
```diff
@@ -1,51 +1,66 @@
...
         {
-            Uuid = uuid;
-            Name = name;
+            Id = id;
+            InteractId = interactId;
+            State = state;
+            // enabling all interact objects. Seems like the default status in the xblock/flat files has it disabled.
+            // TODO: find out where these are actually turned on?
             Type = type;
+
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/InteractObject.cs:22]:
```suggestion
```

**Review by AngeloTadeucci — APPROVED** [review]

File: GameDataParser/Parsers/MapEntityParser.cs (hunk 1 above)
Comment (Sparkymod) [reply GameDataParser/Parsers/MapEntityParser.cs:420]:
```suggestion

```

**Review by Sparkymod — COMMENTED** [review]
Nice

File: MapleServer2/Managers/FieldManager.cs [hunk 3]
```diff
@@ -102,13 +100,7 @@ public FieldManager(int mapId, long instanceId)
...
 
             // Load default InteractObjects
-            List<IFieldObject<InteractObject>> actors = new List<IFieldObject<InteractObject>> { };
-            foreach (MapInteractObject interactObject in MapEntityStorage.GetInteractObject(mapId))
-            {
-                // TODO: Group these fieldActors by their correct packet type.
-                actors.Add(RequestFieldObject(new InteractObject(interactObject.Uuid, interactObject.Name, interactObject.Type) { }));
-            }
-            AddInteractObject(actors);
+
```
Comment (darkvergus) [inline MapleServer2/Managers/FieldManager.cs:103]:
```suggestion
-
```

**Review by darkvergus — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]

File: MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs [hunk 4]
```diff
@@ -33,90 +33,112 @@ public override void Handle(GameSession session, PacketReader packet)
...
+            {
+                int prob = RarityChance[item.Rarity] * masteryDiffFactor / 10000;
+                if (RandomProvider.Get().Next(100) >= prob)
+                {
+                    continue;
+                }
+                for (int i = 0; i < item.Amount; i++)
+                {
+
+                    session.FieldManager.AddItem(session, new Item(item.ItemId)
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs:129]:
```suggestion
                }
                
                for (int i = 0; i < item.Amount; i++)
                {
                    session.FieldManager.AddItem(session, new Item(item.ItemId)
```

File: MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs [hunk 5]
```diff
@@ -33,90 +33,112 @@ public override void Handle(GameSession session, PacketReader packet)
...
-                    return;
-                default:
-                    break;
+                        Rarity = item.Rarity,
+                        Amount = item.Amount
+                    });
+                }
+                numDrop += item.Amount;
+            }
+            if (numDrop > 0)
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs:137]: invert and return

File: MapleServer2/PacketHandlers/Game/MasteryHandler.cs [hunk 6]
```diff
@@ -129,20 +129,21 @@ private static void HandleCraftItem(GameSession session, PacketReader packet)
...
+                Item item = session.Player.Inventory.Items.Values.FirstOrDefault(x => x.Id == ingredient.ItemId && x.Rarity == ingredient.Rarity);
                 if (item == null)
                 {
-                    continue;
+                    return false;
+                }
+
+                if (item.Amount < ingredient.Amount)
+                {
+                    return false;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MasteryHandler.cs:144]: merge into one if statement

File: MapleServer2/PacketHandlers/Game/MasteryHandler.cs [hunk 7]
```diff
@@ -187,20 +188,23 @@ private static bool PlayerHasEnoughMesos(GameSession session, RecipeMetadata rec
...
+                Item item = session.Player.Inventory.Items.Values.FirstOrDefault(x => x.Id == ingredient.ItemId && x.Rarity == ingredient.Rarity);
+                if (item == null)
                 {
-                    return item.Amount >= ingredient.Amount;
+                    return false;
+                }
+
+                if (item.Amount < ingredient.Amount)
+                {
+                    return false;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MasteryHandler.cs:203]: same thing here

**Review by AngeloTadeucci — APPROVED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**